engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/decimal/CeilNode.java

```diff
@@ -1,23 +1,32 @@
 package org.enso.interpreter.node.expression.builtin.number.decimal;
 
 import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
 import com.oracle.truffle.api.profiles.CountingConditionProfile;
 import java.math.BigDecimal;
 import java.math.BigInteger;
 import org.enso.interpreter.dsl.BuiltinMethod;
-import org.enso.interpreter.node.expression.builtin.number.utils.BigIntegerOps;
 import org.enso.interpreter.runtime.number.EnsoBigInteger;
 
 @BuiltinMethod(
     type = "Float",
     name = "ceil",
     description = "Float ceiling, converting to an integer.")
-public class CeilNode extends FloatNode {
+public abstract class CeilNode extends FloatNode {
   private final CountingConditionProfile fitsProfile = CountingConditionProfile.create();
 
-  Object execute(double own) {
+  static CeilNode build() {
+    return CeilNodeGen.create();
+  }
+
+  abstract Object execute(double own);
+
+  @Specialization
+  Object doDouble(double own, @CachedLibrary(limit = "1") InteropLibrary interop) {
```

Comment:
Interesting. I assume the goal here is to delegate to Truffle implementation of `fitsInLong` rather than trying to maintain our own. There should always be _exactly one_ `InteropLibrary` for `double`, thus following might be easier for the system to "swallow":
```suggestion
  @Specialization(limit="1")
  Object doDouble(double own, @CachedLibrary(value = "own") InteropLibrary interop) {
```
(if my code compiles) the peak performance shall be identical, but _getting to the peak performace_ is probably easier in my version (but we don't have a tool to measure it).

Reply:
Thanks for the suggestion, it seems to compile.